backend/handlers.go

```diff
@@ -0,0 +1,626 @@
+					if accept == "application/vnd.ipld.car" {
+						isCar = true
+						break
+					}
+				}
+			}
+		}
+		if !isCar {
+			http.Error(w, "only car format supported", http.StatusBadRequest)
+			return
+		}
+
+		contentPath := ipath.New(r.URL.Path)
+		if contentPath.Mutable() {
+			http.Error(w, "only immutable block requests supported", http.StatusBadRequest)
+			return
+		} else if contentPath.Namespace() != "ipfs" {
+			http.Error(w, "only the ipfs names is supported", http.StatusBadRequest)
+			return
+		}
+
+		carStream, err := simpleSelectorToCar(ctx, bsrv, contentPath.String(), r.URL.Query())
+		if err != nil {
+			http.Error(w, "only the ipfs names is supported", http.StatusBadRequest)
+			return
+		}
+
+		const immutableCacheControl = "public, max-age=29030400, immutable"
+		// immutable! CACHE ALL THE THINGS, FOREVER! wolololol
+		w.Header().Set("Cache-Control", immutableCacheControl)
+
+		// Set modtime to 'zero time' to disable Last-Modified header (superseded by Cache-Control)
+
+		io.Copy(w, carStream)
+		return
+	})
+
+	// Creates metrics handler for total response size. Matches the same metrics
+	// from Kubo:
+	// https://github.com/ipfs/kubo/blob/e550d9e4761ea394357c413c02ade142c0dea88c/core/corehttp/metrics.go#L79-L152
+	sum := prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace:  "ipfs",
+		Subsystem:  "http",
+		Name:       "response_size_bytes",
+		Help:       "The HTTP response sizes in bytes.",
+		Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+	}, nil)
+	err := prometheus.Register(sum)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the HTTP handler for the gateway.
+	handler := promhttp.InstrumentHandlerResponseSize(sum, mux)
+
+	// Add logging
+	handler = withRequestLogger(handler)
+
+	return &http.Server{
+		Handler: handler,
+		Addr:    ":" + strconv.Itoa(port),
+	}, nil
+}
+
+func simpleSelectorToCar(ctx context.Context, bsrv blockservice.BlockService, p string, params url.Values) (io.ReadCloser, error) {
+	pathSegs := strings.Split(p, "/")
+	if len(pathSegs) < 3 || !(pathSegs[0] == "" && pathSegs[1] == "ipfs") {
+		return nil, fmt.Errorf("invalid path")
+	}
+	pathSegs = pathSegs[2:]
+	rootCidStr := pathSegs[0]
+	rootCid, err := cid.Decode(rootCidStr)
+	if err != nil {
+		return nil, err
+	}
+
+	ipfspath, err := path.ParsePath(p)
+	if err != nil {
+		return nil, err
+	}
+
+	r, w := io.Pipe()
+
+	rangeStr, hasRange := params.Get("bytes"), params.Has("bytes")
+	depthStr, hasDepth := params.Get("depth"), params.Has("depth")
+
+	if hasDepth && !(depthStr == "0" || depthStr == "1" || depthStr == "all") {
+		return nil, fmt.Errorf("depth type: %s not supported", depthStr)
+	}
+	var getRange *gateway.GetRange
+	if hasRange {
+		getRange, err = rangeStrToGetRange(rangeStr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	go func() {
+		defer w.Close()
+
+		// Setup header for the output car
+		err = car.WriteHeader(&car.CarHeader{
+			Roots:   []cid.Cid{rootCid},
```

Comment:
It's definitely also `/ipfs/cid/path/to/stuff`. It's unclear to me what is supposed to go in the roots here though. IIUC some of the CAR splitting tools that exist in dag-house/filecoin/etc. for when they want to take a DAG and split it across CAR files put in the root of an incomplete graph at the front of the CAR files.

Maybe we could just remove the roots entirely, but not sure if some tooling expects them to exist. This (perhaps out of date?) document basically leaves all this as 🤷 https://ipld.io/specs/transport/car/carv1/#unresolved-items.

@rvagg what roots are you emitting from Lassie for partial DAGs?

Reply:
Always the requested CID - based on the assumption that since we're working with "verifiable" CARs that you want to start from the root and navigate into your content and that the root is the user's trust anchor that we have to revolve around.

It also gives us a nice workflow for these partial DAGs where we can pipe them into plain UnixFS tooling that assumes the root is where everything starts. So right now you can do things like `lassie fetch -o - <cid>/path/to/thing | car extract <cid>/path/to/thing -` and you have verified download and extraction without needing to do anything else! (Latest `car` can also skip the whole `<cid>/path/to/thing` argument, it'll just fish around in your incomplete DAG and find whatever's in there to extract and lassie will only give you the minimum).

Without additional metadata to embed "/path/to/thing" in the CAR header, I can kind of see an argument for wanting to put an additional "root" for each step of the DAG along that path (which could get quite large!), but that seems like a bad fix for something that's not really essential—we only ever give you just enough to reconstitute what you need to get from the root to your content so I think our tooling should be able to deal with these cases.

Reply:
Oh and I forgot to mention also wrt multiple roots for each step on the path, even if we just only wrote the root that was at the path terminus - we don't have that information when we start writing the CAR and it'd be ideal to not have to buffer and delay until we're deep enough to get that information, because again these path DAGs could get quite large!

And aside, there still remains open the possibility of abusing CARv2 to squish in some kind of metdata if we _really_ want it: https://github.com/ipld/go-car/pull/322, https://github.com/ipld/js-car/pull/89. But I think first we should explore what we're really solving for—if it's just for `ipfs dag import` then let's either (a) fix that, or (b) provide adjacent tooling to make the problems go away. (or more radically, (c) consider that a legacy problem?)